src/FastConv/Convolver.cpp

```diff
@@ -0,0 +1,329 @@
+        m_H_imag[i] = new float[m_blockLengthPlusOne]();
+
+        memset(iFFTTemp, 0, sizeof(float) * doubleBlockLength);
+        memcpy(iFFTTemp, impulseResponse + (i * blockLength), sizeof(float) * std::min(blockLength, m_originIRLengthInSample - i * blockLength));
+
+        m_pCFft->doFft(m_H, iFFTTemp);
+        m_pCFft->splitRealImag(m_H_real[i], m_H_imag[i], m_H);
+    }
+
+    delete[] m_H;
+
+    return Error_t::kNoError;
+}
+
+Error_t UniformlyPartitionedFFTConvolver::reset() {
+    for (int i = 0; i < m_IRNumBlock; ++i) {
+        delete[] m_H_real[i];
+        delete[] m_H_imag[i];
+    }
+    delete[] m_H_real;
+    delete[] m_H_imag;
+    delete[] m_X;
+
+    delete[] aReal;
+    delete[] cReal;
+    delete[] aImag;
+    delete[] cImag;
+
+    delete[] iFFTTemp;
+    delete[] temp;
+
+    delete m_buffer;
+
+    m_pCFft->resetInstance();
+    CFft::destroyInstance(m_pCFft);
+    m_pCFft = nullptr;
+
+    return Error_t::kNoError;
+}
+
+Error_t UniformlyPartitionedFFTConvolver::process(float* output, const float* input, int bufferLength) {
+    int accuLength = 0;
+    int curBufferLength;
+    while (bufferLength) {
+        curBufferLength = std::min(bufferLength, m_blockLength);
+        __processOneBlock(output + accuLength, input + accuLength, curBufferLength);
+        accuLength += curBufferLength;
+        bufferLength -= curBufferLength;
+    }
+    return Error_t::kNoError;
+}
+
+void UniformlyPartitionedFFTConvolver::__processOneBlock(float* output, const float* input, int bufferLength) {
+    assert(bufferLength <= m_blockLength);
+    int doubleBlockLength = m_blockLength << 1;
+    memset(iFFTTemp, 0, sizeof(float) * doubleBlockLength);
+    memcpy(iFFTTemp, input, sizeof(float) * bufferLength);
+    m_pCFft->doFft(m_X_origin, iFFTTemp);
+
+    // do convolution for each block on frequency domain
+    for (int i = 0; i < m_IRNumBlock; ++i) {
+        memcpy(m_X, m_X_origin, sizeof(CFft::complex_t) * doubleBlockLength);
+        __complexVectorMul_I(m_X, m_H_real[i], m_H_imag[i]); // Y[i] = X[i] * H[i], but &Y == &X
+        m_pCFft->doInvFft(iFFTTemp, m_X);
+        CVectorFloat::mulC_I(iFFTTemp, static_cast<float>(doubleBlockLength), doubleBlockLength);
+
+        int offset = i * m_blockLength;
+        for (int j = 0; j < 2; ++j) {
+            __addToRingBuffer(m_buffer->getHead(offset + j * m_blockLength), iFFTTemp + j * m_blockLength, m_blockLength);
+        }
+    }
+
+    // update output using ringbuffer buffer
+    __flushRingBufferToOutput(output, bufferLength);
+
+    applyWetGain(output, bufferLength);
+}
+
+Error_t UniformlyPartitionedFFTConvolver::flushBuffer(float *pfOutputBuffer) {
+    for (int i = 0; i < m_IRNumBlock; ++i) {
+        memcpy(
+            pfOutputBuffer + i * m_blockLength, 
+            m_buffer->getHead(i * m_blockLength), 
+            sizeof(float) * std::min(m_blockLength, m_originIRLengthInSample - 1 - i * m_blockLength)
+        );
```

Comment:
There's a bug here. Denote `numSampleToCopy = sizeof(float) * std::min(m_blockLength, m_originIRLengthInSample - 1 - i * m_blockLength)`, and `bufferHead = m_buffer->getHead(i * m_blockLength)`. Then there's no guarantee that there will always be `numSampleToCopy` samples starting from bufferHead.

Reply:
Fixed: 04995c5937e11713dd3747ea65f37650d2eec19f.